tests/lightning/pytorch/callbacks/test_model_checkpoint.py

```diff
@@ -0,0 +1,228 @@
+
+    def val_dataloader(self):
+        dataset = RandomDataset(32, 16)
+        return torch.utils.data.DataLoader(dataset, batch_size=2)
+
+    def test_dataloader(self):
+        dataset = RandomDataset(32, 16)
+        dl = torch.utils.data.DataLoader(dataset, batch_size=2)
+        self._test_names = ['test_{}_'.format(idx) for idx in range(len(dl))]
+        return dl
+
+    def training_step(self, batch):
+        return self(batch)
+
+    def validation_step(self, batch):
+        loss = self(batch)
+        self.validation_step_outputs.append(loss)
+        return loss
+
+    def test_step(self, batch):
+        loss = self(batch)
+        self.test_step_outputs.append(loss)
+        return loss
+
+    def configure_optimizers(self):
+        return torch.optim.SGD(self.parameters(), lr=1e-3)
+
+    def on_validation_epoch_end(self):
+        self.log("val_loss", torch.stack(self.validation_step_outputs).mean())
+        self.validation_step_outputs.clear()  # free memory
+
+    def configure_model(self):
+        self.module = ExampleModel()
+
+    def set_input_tensor(self, input_tensor: Optional[Tensor]) -> None:
+        pass
+
+    def training_loss_reduction(self) -> MegatronLossReduction:  # noqa: D102
+        # This is the function that takes batch['loss_mask'] and the logits output by the model and reduces the loss
+        return PassThroughLossReduction()
+
+    def validation_loss_reduction(self) -> MegatronLossReduction:  # noqa: D102
+        return PassThroughLossReduction()
+
+
+def setup_test(path, async_save=False):
+    model = ExampleModel()
+
+    data = RandomDataset(32, 64)
+
+    nemo_logger = nl.NeMoLogger(
+        log_dir=path,
+        use_datetime_version=False,
+    )
+
+    strategy = nl.MegatronStrategy(ckpt_async_save=async_save, replace_progress_bar=False)
+
+    trainer = nl.Trainer(
+        max_epochs=5,
+        devices=1,
+        val_check_interval=5,
+        callbacks=nl.ModelCheckpoint(
+            monitor="val_loss",
+            save_top_k=3,
+            save_on_train_epoch_end=True,
+            save_context_on_train_end=False,
+            filename=f'{{step}}-{{epoch}}-{{val_loss}}-{{consumed_samples}}',
+            save_last="link",
+        ),
+        strategy=strategy,
+    )
+    nemo_logger.setup(trainer)
+
+    return data, model, trainer
+
+
+class TestLinkCheckpoint:
+
+    @pytest.mark.unit
+    def test_link_ckpt(self, tmpdir):
+        """Test to ensure that we always keep top_k checkpoints, even after resuming."""
+
+        with reset_megatron_parallel_state():
+            tmp_path = tmpdir / "link_ckpt_test"
+            data, model, trainer = setup_test(tmp_path, async_save=False)
+
+            trainer.fit(model, data)
+
+            checkpoint_dir = Path(tmp_path / "default" / "checkpoints")
+            dist_checkpoints = [d for d in list(checkpoint_dir.glob("*")) if d.is_dir()]
+            last_checkpoints = [d for d in dist_checkpoints if d.match("*last")]
```

Comment:
Thank you @ashors1 for adding the test. I think this does the job. Can we also add an assert line ensuring there are only `save_top_k` number of checkpoints in the dir.

Reply:
Yep, done